Compile jemalloc from source.

pgroudas opened the pull request:
### Fix Makefile

I am unsure as to why, but the dependency expressed as "mcrouter: deps"
doesn't seem to work correctly.

When invoking the equivalent of "make mcrouter", I expect it to build
all the transitive dependencies of "deps", but it only builds the first
dependency that it determines it needs to build, and then exits with
a return value of "2".

I'm just reverting this to the previous style that is redundant but
works correctly.

### "git clean" during checkout of dependency

This repo defines "recipes" for dependencies which are essentially build
scripts.  Some of the build systems used (or at least the way they are
used) will cache some state between invocations, which is confusing when
testing different values for things like "LDFLAGS" and
"LD_LIBRARY_PATH".  Specifically, the cmake based build scripts.

This change ensures that when force re-running a given recipe, we do a
full clean of the repo.

n.b.: we are *not* cleaning the target install dir.

### Build jemalloc from source.

This introduces "jemalloc" as a library dependency that we build from
source.

While this library is available from our runtime system's package
respository, mcrouter would log an error at startup implying that there
is a page size mis-match between the current kernel, and the system on
which jemalloc was compiled.

As this package available via the EPEL[1] and not Amazon Linux directly,
this isn't entirely surprising.

Compiling "jemalloc" is very straightforward, but getting all the
dependencies to correctly link against it during their builds was a bit
challenging.  Ultimately I made the following changes:

1. The default value we set for *all* recipes is:

        LDFLAGS="-L$INSTALL_DIR/lib -ljemalloc

- "-ljemalloc" means "link against the 'jemalloc'" library.
- "-L$INSTALL_DIR/lib" search "$INSTALL_DIR/lib" when looking for
libraries.

2. I removed redundant LDFLAGS from each recipe following the change to
the default.

3. Some builds, notable those that produce binary executables (not just
shared libraries), also need to include "-Wl,-rpath=$INSTALL_DIR/lib",
which embeds metadata about the dynamic libraries' location into the
executable.  Conceptually, this doesn't make sense to me, as we are
expecting users to manage their own library path, but not setting this
causes the builds to fail, and I have a limited appetite for debugging
and understanding these build tools.

[1]: https://fedoraproject.org/wiki/EPEL

[Clubhouse Story pgroudas/compile-jemalloc-from-source](https://app.clubhouse.io/internal/story/pgroudas/compile-jemalloc-from-source)

Risk is low

favila reviewed (approved):
Is it yet confirmed that this fixes the issue?

I'm not understanding why we're getting this error in the first place *because* we're building and running on the same platform against a library amazon built themselves. Is there any answer to that mystery?

pgroudas commented:
> Is it yet confirmed that this fixes the issue?

It is not.  As the previous build had `LDFLAGS="-ljemalloc"` I think all of the dependencies that supported it were linking against the system version, which was very old.

I am still working through the changes necessary to update the build flags for everything to just get it all working from top to bottom to see if this makes the error go away.

> I'm not understanding why we're getting this error in the first place because we're building and running on the same platform against a library amazon built themselves. Is there any answer to that mystery?

I will point out that the version available in Amazon's repository is 3.6.0 which is from April 2015.  This particular package comes from the EPEL repos, so I don't think it is actually built by Amazon.

I also think it is poor design to have the compile-time kernel page size be required to be used at runtime, but it is what it is.

Commit:
Fix Makefile

I am unsure as to why, but the dependency expressed as "mcrouter: deps"
doesn't seem to work correctly.

When invoking the equivalent of "make mcrouter", I expect it to build
all the transitive dependencies of "deps", but it only builds the first
dependency that it determines it needs to build, and then exits with
a return value of "2".

I'm just reverting this to the previous style that is redundant but
works correctly.

Commit:
"git clean" during checkout of dependency

This repo defines "recipes" for dependencies which are essentially build
scripts.  Some of the build systems used (or at least the way they are
used) will cache some state between invocations, which is confusing when
testing different values for things like "LDFLAGS" and
"LD_LIBRARY_PATH".  Specifically, the cmake based build scripts.

This change ensures that when force re-running a given recipe, we do a
full clean of the repo.

n.b.: we are *not* cleaning the target install dir.

Commit:
Build jemalloc from source.

This introduces "jemalloc" as a library dependency that we build from
source.

While this library is available from our runtime system's package
respository, mcrouter would log an error at startup implying that there
is a page size mis-match between the current kernel, and the system on
which jemalloc was compiled.

As this package available via the EPEL[1] and not Amazon Linux directly,
this isn't entirely surprising.

Compiling "jemalloc" is very straightforward, but getting all the
dependencies to correctly link against it during their builds was a bit
challenging.  Ultimately I made the following changes:

1. The default value we set for *all* recipes is:

        LDFLAGS="-L$INSTALL_DIR/lib -ljemalloc

- "-ljemalloc" means "link against the 'jemalloc'" library.
- "-L$INSTALL_DIR/lib" search "$INSTALL_DIR/lib" when looking for
libraries.

2. I removed redundant LDFLAGS from each recipe following the change to
the default.

3. Some builds, notable those that produce binary executables (not just
shared libraries), also need to include "-Wl,-rpath=$INSTALL_DIR/lib",
which embeds metadata about the dynamic libraries' location into the
executable.  Conceptually, this doesn't make sense to me, as we are
expecting users to manage their own library path, but not setting this
causes the builds to fail, and I have a limited appetite for debugging
and understanding these build tools.

[1]: https://fedoraproject.org/wiki/EPEL

pgroudas commented:
@favila Please take another look.

I have now confirmed that this does in fact prevent that error / warning at startup:

```
$ LD_LIBRARY_PATH=/home/clubhouse-admin/mcrouter-install/install/lib ./mcrouter --config-str='{"pools":{"A":{"servers":["127.0.0.1:5001"]}},
                  "route":"PoolRoute|A"}' -p 22122
I0503 20:02:35.713507 30277 StandaloneUtils.cpp:479] .0 mcrouter startup (30277)
I0503 20:02:35.713589 30277 StandaloneUtils.cpp:192] Starting Memcache router
I0503 20:02:35.713616 30277 Server-inl.h:493] Spawning AsyncMcServer
I0503 20:02:35.714282 30277 FunctionScheduler.cpp:397] Starting FunctionScheduler with 0 functions.
I0503 20:02:35.714722 30277 CarbonRouterInstance-inl.h:516] creating config builder
I0503 20:02:35.715301 30277 CarbonRouterInstance-inl.h:486] started reconfiguring
I0503 20:02:35.715438 30277 CarbonRouterInstance-inl.h:504] reconfigured 1 proxies with 1 pools, 1 clients 8fbe416445f80756f529cbd598f5d387)
W0503 20:02:35.715590 30277 Server-inl.h:56] acl checker will not be enabled.
```

pgroudas commented:
I'm happy to address any feedback in a follow-up PR, but I'm going to merge this as is so I can get the builds started, as I'll need to reference the final git sha
